src/nodes/Collection.ts

```diff
@@ -207,4 +214,90 @@ export abstract class Collection extends NodeBase {
         )
     }
   }
+
+  /** Merge another collection into this */
+  merge(source: Collection|Document): void {
+    let sourceNode: Node;
+
+    if (source instanceof Document) {
+      if (isNode(source.contents)) {
+        sourceNode = source.contents
+      } else {
+        throw new Error(`Cannot merge Document content type "${typeof source.contents}" into Document`)
+      }
+    }  else if (isNode(source)) {
+      sourceNode = source
+    } else {
+      throw new Error(`Cannot merge node type "${typeof source}" into Document`)
+    }
+
+    function getPath(ancestry: readonly (Document | Node | Pair)[]): (string|number)[] {
+      return ancestry.reduce<(string|number)[]>((p, { key }: any) => {
+        return (key !== undefined) ? [...p, key.value] : p;
+      }, []);
+    }
+
+    visit(sourceNode, {
+      Map: (_, node, ancestors) => {
+        const path = getPath(ancestors);
+        // In case both the source and the target node have comments or spaces
+        // We'll move them to their first child so they do not conflict
+        if(this.hasIn(path)) {
+          const targetNode = this.getIn(path, true)
+
+          if (!isMap(targetNode)) {
+            const type = (<any> targetNode).constructor?.name || typeof targetNode;
+            throw new Error(`Conflict at "${path.join('.')}": Destination node is of type ${type}, the node to be merged is of type ${node.constructor.name}`)
+          }
+
+          targetNode.moveCommentsAndSpaceToFirstChildNode()
+          node.moveCommentsAndSpaceToFirstChildNode()
+        }
+      },
+      Pair: (_, node, ancestors) => {
+        if (isScalar(node.value)) {
+          const path = getPath(ancestors);
+          if (this.hasIn([...path, (<Scalar> node.key).value])) {
+            this.setIn(path, node);
+          } else {
+            this.addIn(path, node);
+          }
+          return visit.SKIP;
+        }
+      },
+      Seq: (_, node, ancestors) => {
+        const path = getPath(ancestors);
+        if(this.hasIn(path)) {
+          const targetNode = this.getIn(path, true)
+          if (!isSeq(targetNode)) {
+            const type = (<any> targetNode).constructor?.name || typeof targetNode;
+            throw new Error(`Conflict at "${path.join('.')}": Destination node is of type ${type}, the node to be merged is of type ${node.constructor.name}`)
+          }
+          targetNode.moveCommentsAndSpaceToFirstChildNode()
+          node.moveCommentsAndSpaceToFirstChildNode()
+        }
+        node.items
+          .forEach((item) => {
+            this.addIn(path, item);
+          });
+        return visit.SKIP;
+      },
+    });
+  }
+
+  moveCommentsAndSpaceToFirstChildNode(): void {
+    const firstChildNode = this.getFirstChildNode()
```

Comment:
I was *very* torn on this. Since `YAMLSeq` and `YAMLMap` differ slightly in how that first Node is obtained, I didn't want to start guessing in the Collection method: If someone `extend`ed the Collection with a slightly different class (eg. while writing a custom tag), this could (silently) cause unexpected behaviour when merging, so I just gave the extension the job to specify how it's done.
But I do agree that abstract methods should be used sparingly, so it's your call.